### PR TITLE
[git-sync] Allow random failures

### PR DESCRIPTION
**Pull request by aledbf**
This PR introduces a flag to allow failures in the sync operation

```
$ docker run -it \
> -e GIT_SYNC_REPO="https://********** repo url **********.git" \
> -e GIT_SYNC_DEST="/git" \
> -e GIT_SYNC_MAX_SYNC_FAILURES="10" \
> gcr.io/google_containers/git-sync
2016/04/21 12:52:17 setting up the git credential cache
2016/04/21 12:52:18 clone "https://********** repo url **********.git": Cloning into '/git'...
2016/04/21 12:52:19 fetch "master": From https://********** repo url **********
 * branch            master     -> FETCH_HEAD
Already up-to-date.
2016/04/21 12:52:19 reset "HEAD": HEAD is now at befa882 feat(*): incorporación de versión 06958961
2016/04/21 12:52:19 wait 10 seconds
2016/04/21 12:52:29 done
2016/04/21 12:52:30 fetch "master": From https://********** repo url **********
 * branch            master     -> FETCH_HEAD
Already up-to-date.
2016/04/21 12:52:30 reset "HEAD": HEAD is now at befa882 feat(*): incorporación de versión 06958961
2016/04/21 12:52:30 wait 10 seconds
2016/04/21 12:52:40 done
```

after disconnecting internet access

```
2016/04/21 12:53:20 unexpected error syncing repo: error running command "git pull origin master" : exit status 1: fatal: unable to access 'https://********** repo url **********.git/': Could not resolve host: ********** host **********
2016/04/21 12:53:20 wait 10 seconds before retryng
2016/04/21 12:54:10 unexpected error syncing repo: error running command "git pull origin master" : exit status 1: fatal: unable to access 'https://********** repo url **********.git/': Could not resolve host: ********** host **********
2016/04/21 12:54:10 wait 10 seconds before retryng
2016/04/21 12:55:00 unexpected error syncing repo: error running command "git pull origin master" : exit status 1: fatal: unable to access 'https://********** repo url **********.git/': Could not resolve host: ********** host **********
2016/04/21 12:55:00 wait 10 seconds before retryng
2016/04/21 12:55:50 unexpected error syncing repo: error running command "git pull origin master" : exit status 1: fatal: unable to access 'https://********** repo url **********.git/': Could not resolve host: ********** host **********
2016/04/21 12:55:50 wait 10 seconds before retryng
2016/04/21 12:56:40 unexpected error syncing repo: error running command "git pull origin master" : exit status 1: fatal: unable to access 'https://********** repo url **********.git/': Could not resolve host: ********** host **********
2016/04/21 12:56:40 wait 10 seconds before retryng
2016/04/21 12:57:30 error syncing repo: error running command "git pull origin master" : exit status 1: fatal: unable to access 'https://********** repo url **********.git/': Could not resolve host: ********** host **********
```

### CODE REVIEW

**Comment by aledbf** [comment]
@bprashanth who can review this?

File: git-sync/main.go [hunk 1]
```diff
@@ -40,6 +40,9 @@ var flWait = flag.Int("wait", envInt("GIT_SYNC_WAIT", 0), "number of seconds to
 var flOneTime = flag.Bool("one-time", envBool("GIT_SYNC_ONE_TIME", false), "exit after the initial checkout")
 var flDepth = flag.Int("depth", envInt("GIT_SYNC_DEPTH", 0), "shallow clone with a history truncated to the specified number of commits")
 
+var flAllowFail = flag.Bool("allow-failures", envBool("GIT_SYNC_ALLOW_FAILURES", false), "allow temporal git failures (the first pull must succeed)")
```
Comment (thockin) [inline git-sync/main.go]: Do we need two flags?  why not just --max-sync-failures=0 means none allowed, and any other value means they are allowed.

I'm a fan of very specific flag names, `max-sync-failures` ?

File: git-sync/main.go [hunk 2]
```diff
@@ -40,6 +40,9 @@ var flWait = flag.Int("wait", envInt("GIT_SYNC_WAIT", 0), "number of seconds to
...
+var flFailCount = flag.Int("failure-count", envInt("GIT_SYNC_FAILURE_COUNT", 1), "consecutive failures allowed")
```
Comment (thockin) [inline git-sync/main.go]: "...failures allowed before aborting" ?

File: git-sync/main.go [hunk 3]
```diff
@@ -95,11 +98,24 @@ func main() {
...
 	for {
 		if err := syncRepo(*flRepo, *flDest, *flBranch, *flRev, *flDepth); err != nil {
-			log.Fatalf("error syncing repo: %v", err)
+			if initial || !*flAllowFail || (*flAllowFail && failCount > *flFailCount) {
+				log.Fatalf("error syncing repo: %v", err)
+			}
+
+			failCount++
+			log.Printf("unexpected error syncing repo: %v", err)
+			log.Printf("wait %d seconds before retryng", *flWait)
```
Comment (thockin) [inline git-sync/main.go]: "waiting"

File: git-sync/main.go [hunk 4]
```diff
@@ -95,11 +98,24 @@ func main() {
 		}
 	}
 
+	initial := true
```
Comment (thockin) [inline git-sync/main.go]: initial what?  initialSync or firstSync or something, please

**Comment by aledbf** [comment]
ping @thockin

File: git-sync/main.go [hunk 5]
```diff
@@ -95,16 +98,29 @@ func main() {
 		}
 	}
 
+	initialSync := true
+	failCount := 0
 	for {
 		if err := syncRepo(*flRepo, *flDest, *flBranch, *flRev, *flDepth); err != nil {
-			log.Fatalf("error syncing repo: %v", err)
+			if initialSync || *flMaxSyncFailures == 0 || (*flMaxSyncFailures > 0 && failCount >= *flMaxSyncFailures) {
```
Comment (eparis) [inline git-sync/main.go]: Is `-1` supposed to mean 'forever' ?

Isn't `*flMaxSyncFailures == 0 || (*flMaxSyncFailures > 0 && failCount >= *flMaxSyncFailures)` just: `failCount >= *flMaxSyncFailures` ?

Comment (aledbf) [inline git-sync/main.go]: > Is -1 supposed to mean 'forever' ?

No but I can add that condition

Comment (aledbf) [inline git-sync/main.go]: > failCount >= *flMaxSyncFailures ?

done

**Pull request merged**